src/sciline/_provider.py

```diff
@@ -259,7 +274,46 @@ def _module_name(x: Any) -> str:
     return getattr(inspect.getmodule(x), '__name__', '')
 
 
-def _get_provider_args(func: ToProvider) -> inspect.FullArgSpec:
+def _unwrap_decorated(func: ToProvider) -> ToProvider:
     if (wrapped := getattr(func, '__wrapped__', None)) is not None:
-        return _get_provider_args(wrapped)
-    return inspect.getfullargspec(func)
+        return _unwrap_decorated(wrapped)
+    return func
+
+
+def _get_args_and_types(
+    func: ToProvider,
+) -> tuple[list[str], list[str], dict[str, Any]]:
+    if inspect.isfunction(func):
+        return _get_func_args_and_types(func)
+    if inspect.ismethod(func):
+        return _get_method_args_and_types(func)
+    if isinstance(func, type):
+        raise TypeError(
+            "Classes are not supported as providers because their type annotations "
+            f"cannot reliably be extracted. Got class {func}. "
+            "Consider using a free function, classmethod, or staticmethod."
```

Comment:
Is it worth mentioning that this is about classes, not class instances? While clear to us, an average user may forget to instantiate a class and not know the difference?

Reply:
Better?